File: apps/api/src/routes/incidents.ts
import { Router } from "express";
import type { Response as ExpressResponse } from "express";
import { authMiddleware } from "../middleware/auth.js";
import { requireAdmin } from "../middleware/permissions.js";
import { respondServiceError as respondIncidentServiceError } from "../lib/respond-service-error.js";
import {
  listIncidents,
  createIncident,
  getIncidentUsage,
  listIncidentImages,
  createIncidentImage,
  deleteIncidentImage,
  updateIncident,
  deleteIncident,
} from "../services/incident-service.js";

const router = Router();
router.use(authMiddleware);

const respondIncidentBadRequest = (res: ExpressResponse, message: string): void => {
  res.status(400).json({ success: false, error: message });
};

// GET /incidents
router.get("/", requireAdmin(), async (_req, res) => {
  try {
    const incidents = await listIncidents();
    res.json({ success: true, data: incidents });
  } catch (err) {
    console.error("❌ [INCIDENTS] GET:", err);
    res.status(500).json({ success: false, error: "Erro interno ao listar incidentes" });
  }
});

// POST /incidents
router.post("/", requireAdmin(), async (req, res) => {
  const { name, color } = req.body as { name?: string; color?: string };
  if (!name || name.trim().length < 2) {
    respondIncidentBadRequest(res, "Nome do incidente é obrigatório e deve ter pelo menos 2 caracteres.");
    return;
  }
  try {
    const result = await createIncident({ name, color });
    if (!result.ok) {
      respondIncidentServiceError(res, result, "Erro interno ao criar incidente");
      return;
    }
    const { data } = result;
    res.json({ success: true, data });
  } catch (err) {
    console.error("❌ [INCIDENTS] POST:", err);
    res.status(500).json({ success: false, error: "Erro interno ao criar incidente" });
  }
});

// PUT /incidents
router.put("/", requireAdmin(), async (req, res) => {
  const { id, name, color } = req.body as { id?: string; name?: string; color?: string };
  if (!id || !name || name.trim().length < 2) {
    respondIncidentBadRequest(res, "ID e nome do incidente são obrigatórios.");
    return;
  }
  try {
    const result = await updateIncident({ id, name, color });
    if (!result.ok) {
      respondIncidentServiceError(res, result, "Erro interno ao atualizar incidente");
      return;
    }
    res.json({ success: true, message: "Incidente atualizado com sucesso" });
  } catch (err) {
    console.error("❌ [INCIDENTS] PUT:", err);
    res.status(500).json({ success: false, error: "Erro interno ao atualizar incidente" });
  }
});

// DELETE /incidents?id=
router.delete("/", requireAdmin(), async (req, res) => {
  const { id } = req.query as Record<string, string>;
  if (!id) {
    respondIncidentBadRequest(res, "ID do incidente é obrigatório.");
    return;
  }
  try {
    const result = await deleteIncident(id);
    if (!result.ok) {
      respondIncidentServiceError(res, result, "Erro interno ao excluir incidente");
      return;
    }
    res.json({ success: true, data: { success: true } });
  } catch (err) {
    console.error("❌ [INCIDENTS] DELETE:", err);
    if (err instanceof Error && err.message.includes("foreign key")) {
      respondIncidentBadRequest(res, "Este incidente está sendo usado em outros registros e não pode ser excluído.");
      return;
    }
    res.status(500).json({ success: false, error: "Erro interno ao excluir incidente" });
  }
});

// GET /incidents/usage?incidentId=
router.get("/usage", requireAdmin(), async (req, res) => {
  try {
    const incidentId = typeof req.query.incidentId === "string" ? req.query.incidentId : undefined;
    if (!incidentId) { respondIncidentBadRequest(res, "ID do incidente é obrigatório."); return; }
    const result = await getIncidentUsage(incidentId);

    res.json({
      success: true,
      data: {
        inUse: result.data.inUse,
        usageCount: result.data.usageCount,
        usageDetails: result.data.usageDetails,
      },
    });
  } catch (err) {
    console.error("❌ [INCIDENTS] usage GET:", err);
    res.status(500).json({ success: false, error: "Erro interno" });
  }
});

// GET /api/incidents/images
router.get("/images", requireAdmin(), async (_req, res) => {
  try {
    const result = await listIncidentImages();
    res.json({ success: true, data: { items: result.items } });
  } catch (err) {
    console.error("❌ [API_INCIDENTS] GET /images:", err);
    res.status(500).json({ success: false, error: "Erro ao listar imagens" });
  }
});

// POST /api/incidents/images
router.post("/images", requireAdmin(), async (req, res) => {
  try {
    const body = req.body as Record<string, unknown>;
    const base64 = typeof body.image === "string" ? body.image : null;
    const filename = typeof body.filename === "string" ? body.filename : null;
    if (!base64 || !filename) {
      respondIncidentBadRequest(res, "Dados inválidos");
      return;
    }
    const result = await createIncidentImage({ image: base64, filename });
    if (!result.ok) {
      respondIncidentServiceError(res, result, "Erro ao salvar imagem");
      return;
    }
    res.json({ success: true, data: { filename: result.data.filename, url: result.data.url } });
  } catch (err) {
    console.error("❌ [API_INCIDENTS] POST /images:", err);
    res.status(500).json({ success: false, error: "Erro ao salvar imagem" });
  }
});

// DELETE /api/incidents/images?filename=
router.delete("/images", requireAdmin(), async (req, res) => {
  try {
    const filename = req.query.filename as string | undefined;
    if (!filename) {
      respondIncidentBadRequest(res, "Nome de arquivo inválido");
      return;
    }
    const result = await deleteIncidentImage(filename);
    if (!result.ok) {
      respondIncidentServiceError(res, result, "Erro ao excluir imagem");
      return;
    }
    res.json({ success: true, message: "Imagem excluída com sucesso" });
  } catch (err) {
    console.error("❌ [API_INCIDENTS] DELETE /images:", err);
    res.status(500).json({ success: false, error: "Erro ao excluir imagem" });
  }
});

export default router;

File: apps/backend/src/silo/api/routers/incidents.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy import and_, delete, insert, select, update
from sqlalchemy.engine import Connection

from silo.api.dependencies import get_db, require_admin
from silo.api.responses import build_success_payload
from silo.db.models import legacy_tables
from silo.db.serialization import serialize_legacy_row
from silo.services.common import is_service_error, service_error_response, service_failure, service_success
from silo.storage.uploads import decode_base64_data_uri, delete_upload_file, is_safe_filename, list_upload_files, store_buffer_as_webp

router = APIRouter(prefix="/api/incidents", tags=["incidents"])

NO_INCIDENTS_CATEGORY_ID = "no-incidents"


@router.get("")
@router.get("/")
async def list_incidents(
    _current_user: object = Depends(require_admin),
    db: Connection = Depends(get_db),
):
    items = _list_incidents(db)
    return build_success_payload(items)


@router.post("")
@router.post("/")
async def create_incident(
    payload: dict[str, object],
    _current_user: object = Depends(require_admin),
    db: Connection = Depends(get_db),
):
    result = _create_incident(db, payload)
    if is_service_error(result):
        response = service_error_response(result, "Erro interno ao criar incidente")
        assert response is not None
        return response
    return build_success_payload(result["data"])


@router.put("")
@router.put("/")
async def update_incident(
    payload: dict[str, object],
    _current_user: object = Depends(require_admin),
    db: Connection = Depends(get_db),
):
    result = _update_incident(db, payload)
    if is_service_error(result):
        response = service_error_response(result, "Erro interno ao atualizar incidente")
        assert response is not None
        return response
    return build_success_payload(message="Incidente atualizado com sucesso")


@router.delete("")
@router.delete("/")
async def delete_incident(
    id: str | None = Query(default=None),
    _current_user: object = Depends(require_admin),
    db: Connection = Depends(get_db),
):
    if not id:
        return service_error_response(service_failure("ID do incidente é obrigatório.", 400), "Erro interno ao excluir incidente")

    result = _delete_incident(db, id)
    if is_service_error(result):
        response = service_error_response(result, "Erro interno ao excluir incidente")
        assert response is not None
        return response
    return build_success_payload({"success": True})


@router.get("/usage")
async def get_usage(
    incidentId: str | None = Query(default=None),
    _current_user: object = Depends(require_admin),
    db: Connection = Depends(get_db),
):
    if not incidentId:
        return service_error_response(service_failure("ID do incidente é obrigatório.", 400), "Erro interno")

    result = _get_incident_usage(db, incidentId)
    if is_service_error(result):
        response = service_error_response(result, "Erro interno")
        assert response is not None
        return response

    return build_success_payload(result["data"])


@router.get("/images")
async def list_images(
    _current_user: object = Depends(require_admin),
):
    items = list_upload_files("incidents")
    return build_success_payload({"items": items})


@router.post("/images")
async def create_image(
    payload: dict[str, object],
    _current_user: object = Depends(require_admin),
):
    image = payload.get("image")
    filename = payload.get("filename")
    if not isinstance(image, str) or not isinstance(filename, str):
        return service_error_response(service_failure("Dados inválidos", 400), "Erro ao salvar imagem")

    buffer = decode_base64_data_uri(image)
    stored = store_buffer_as_webp("incidents", filename, buffer)
    if isinstance(stored, dict):
        return service_error_response(service_failure(stored["error"], 400), "Erro ao salvar imagem")

    return build_success_payload({"filename": stored.filename, "url": stored.url})


@router.delete("/images")
async def delete_image(
    filename: str | None = Query(default=None),
    _current_user: object = Depends(require_admin),
):
    if not filename:
        return service_error_response(service_failure("Nome de arquivo inválido", 400), "Erro ao excluir imagem")
    result = _delete_incident_image(filename)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao excluir imagem")
        assert response is not None
        return response
    return build_success_payload(message="Imagem excluída com sucesso")


def _list_incidents(db: Connection) -> list[dict[str, object]]:
    category_table = legacy_tables["product_problem_category"]
    rows = db.execute(
        select(category_table)
        .where(category_table.c.id != NO_INCIDENTS_CATEGORY_ID)
        .order_by(category_table.c.sort_order.asc(), category_table.c.name.asc())
    ).mappings().all()
    return [serialize_legacy_row(row) for row in rows]


def _create_incident(db: Connection, payload: dict[str, object]) -> dict[str, object]:
    category_table = legacy_tables["product_problem_category"]

    name = _optional_str(payload.get("name"))
    color = _optional_str(payload.get("color"))
    if not name or len(name.strip()) < 2:
        return service_failure("Nome do incidente é obrigatório e deve ter pelo menos 2 caracteres.", 400)

    normalized_name = name.strip()
    existing = db.execute(
        select(category_table.c.id).where(category_table.c.name == normalized_name).limit(1)
    ).first()
    if existing is not None:
        return service_failure("Nome de incidente já existe.", 400)

    new_incident = {
        "id": _new_uuid(),
        "name": normalized_name,
        "color": color or "#6B7280",
        "is_system": False,
        "sort_order": 999,
    }
    db.execute(insert(category_table).values(new_incident))
    db.commit()
    return service_success(serialize_legacy_row(new_incident))


def _update_incident(db: Connection, payload: dict[str, object]) -> dict[str, object]:
    category_table = legacy_tables["product_problem_category"]

    incident_id = _optional_str(payload.get("id"))
    name = _optional_str(payload.get("name"))
    color = _optional_str(payload.get("color"))
    if not incident_id or not name or len(name.strip()) < 2:
        return service_failure("ID e nome do incidente são obrigatórios.", 400)
    if incident_id == NO_INCIDENTS_CATEGORY_ID:
        return service_failure("Não é possível editar esta categoria.", 400)

    normalized_name = name.strip()
    existing = db.execute(
        select(category_table.c.id)
        .where(and_(category_table.c.name == normalized_name, category_table.c.id != incident_id))
        .limit(1)
    ).first()
    if existing is not None:
        return service_failure("Nome de incidente já existe.", 400)

    db.execute(
        update(category_table)
        .where(category_table.c.id == incident_id)
        .values(name=normalized_name, color=color or "#6B7280", updated_at=_now_naive())
    )
    db.commit()
    return service_success(None)


def _delete_incident(db: Connection, incident_id: str) -> dict[str, object]:
    category_table = legacy_tables["product_problem_category"]
    activity_table = legacy_tables["product_activity"]
    problem_table = legacy_tables["product_problem"]

    category = db.execute(
        select(category_table.c.id, category_table.c.is_system, category_table.c.name)
        .where(category_table.c.id == incident_id)
        .limit(1)
    ).first()
    if category is None:
        return service_failure("Incidente não encontrado.", 404)
    if bool(category.is_system):
        return service_failure(f'"{category.name}" é uma categoria do sistema e não pode ser excluída.', 400)

    usage_in_activities = db.execute(
        select(activity_table.c.id).where(activity_table.c.problem_category_id == incident_id)
    ).all()
    usage_in_problems = db.execute(
        select(problem_table.c.id).where(problem_table.c.problem_category_id == incident_id)
    ).all()
    total_usage = len(usage_in_activities) + len(usage_in_problems)
    if total_usage > 0:
        if total_usage == 1:
            return service_failure("Este incidente está sendo usado em 1 registro e não pode ser excluído.", 400)
        return service_failure(
            f"Este incidente está sendo usado em {total_usage} registros e não pode ser excluído.",
            400,
        )

    db.execute(delete(category_table).where(category_table.c.id == incident_id))
    db.commit()
    return service_success(None)


def _get_incident_usage(db: Connection, incident_id: str) -> dict[str, object]:
    activity_table = legacy_tables["product_activity"]
    problem_table = legacy_tables["product_problem"]

    activities = db.execute(
        select(activity_table.c.id).where(activity_table.c.problem_category_id == incident_id)
    ).all()
    problems = db.execute(
        select(problem_table.c.id).where(problem_table.c.problem_category_id == incident_id)
    ).all()
    total_usage = len(activities) + len(problems)

    return service_success(
        {
            "inUse": total_usage > 0,
            "usageCount": total_usage,
            "usageDetails": {
                "activities": len(activities),
                "problems": len(problems),
            },
        }
    )


def _delete_incident_image(filename: str) -> dict[str, object]:
    if not is_safe_filename(filename):
        return service_failure("Nome de arquivo inválido", 400)
    delete_upload_file("incidents", filename)
    return service_success(None)


def _optional_str(value: object | None) -> str | None:
    return value if isinstance(value, str) else None


def _new_uuid() -> str:
    import uuid

    return str(uuid.uuid4())


def _now_naive():
    from datetime import datetime

    return datetime.now()
